clippy_lints/src/uninhabited_reference.rs

```diff
@@ -0,0 +1,81 @@
+use clippy_utils::diagnostics::span_lint;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, ExprKind, FnDecl, FnRetTy, TyKind, UnOp};
+use rustc_hir_analysis::hir_ty_to_ty;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// It detects references to uninhabited types, such as `!` and
+    /// warns when those are either dereferenced or returned from a function.
+    ///
+    /// ### Why is this bad?
+    /// Dereferencing a reference to an uninhabited type would create
+    /// an instance of such a type, which cannot exist. This constitutes
+    /// undefined behaviour. Such a reference could have been created
+    /// by `unsafe` code.
+    ///
+    /// ### Example
+    /// The following function can return a reference to an uninhabited type
+    /// (`Infallible`) because it uses `unsafe` code to create it. However,
+    /// the user of such a function could dereference the return value and
+    /// trigger an undefined behaviour from safe code.
+    ///
+    /// ```no_run
+    /// fn create_ref() -> &'static std::convert::Infallible {
+    ///     unsafe { std::mem::transmute(&()) }
+    /// }
+    /// ```
+    #[clippy::version = "1.76.0"]
+    pub UNINHABITED_REFERENCE,
+    suspicious,
+    "reference to uninhabited type"
+}
+
+declare_lint_pass!(UninhabitedReference => [UNINHABITED_REFERENCE]);
+
+impl LateLintPass<'_> for UninhabitedReference {
+    fn check_expr_post(&mut self, cx: &LateContext<'_>, expr: &'_ rustc_hir::Expr<'_>) {
+        if expr.span.from_expansion() {
```

Comment:
I am afraid we would get a very large number of warnings. For example,

```rust
#[derive(Debug)]
enum Empty {}
```

would trigger the lint, as the derived `Debug` implementation will use a `&self` parameter, and `Self` is uninhabited.

Reply:
Does `Debug` dereference the `Empty`?  IIUC your lint implementation only checks the return type.

Also trying to debug print this also leads to a `SIGILL`: [`Playground`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2a25ab8e116bbcc7c651bda13f2a5bcb)

Reply:
The lint implementation checks the return type indeed, the warning is because `Debug` dereferences the content. The expansion looks like:

```rust
impl ::core::fmt::Debug for Debug {
    #[inline]
    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
        match *self {}
    }
}
```

In the top 500 crates, we get 63 dereferences of an uninhabited type. I'll check how many would return a reference to one and report here.

Reply:
I'm also doing some experiments right now. On the one hand, you probably shouldn't derive `Debug` on that type. OTOH it is a lot of noise.

Reply:
When linting in non-external-macros, the only additional crate that triggers the lint is `petgraph`, which implements `Clone` for an uninhabited type, because that type is used as a marker type and part of a struct as a `PhantomType`  that derives `Clone`.

So we shouldn't line external macros, as the derives can't be just removed from marker types (which is the main use case that I saw).

---

Besides macros the lint still lints on `powerfmt-0.2.0` and `tonic-0.10.2`. On `tonic` it only triggers the lint, because the one and only enum variant is `cfg`ed out. `powerfmt` seems legit (or at least something one should `#[allow]`).

Reply:
So I think we can lint in macros, just not external macros

Reply:
I've made the change. I'll also propose a patch to `petgraph`, as the `copyclone!` macro (which is not an external macro when linting `petgraph` itself) can be replaced by an automatic `Clone` derivation which would not lint.